### PR TITLE
feat(woodpecker-server)!: add monitoring resources and other improvements

### LINKED ISSUE #21: Prometheus Operator CRDs - ServiceMonitor
it would be nice, that this helmcharts deploy also a ServiceMonitor, which select all Service (and there behind all Pods) to scrape all Metrics.

https://prometheus-operator.dev/docs/operator/api/#monitoring.coreos.com/v1.ServiceMonitor

### PR DESCRIPTION
**Pull request by genofire**
@pat-s please review this time faster.

https://github.com/woodpecker-ci/helm/pull/22

fix #21

---
some things already merged by #77 

---
### BREAKING CHANGES

- split `metrics.enabled` and the PodMonitor deployment to `prometheus.podmonitor.enabled`
  - move `metrics.interval` to `prometheus.podmonitor.interval`

### Features
- metrics port also on services
- `prometheus.podmonitor.interval` optional (for use global server value)
- `prometheus.podmonitor.label` for dedicated labels only on pod-monitor
- first prometheus rule with alerting (is there a worker)
- prepare to ship grafana-dashboard

### Bugfixes
- allow usage of `extraVolumeMounts` without `persistenceVolume.enabled` (#97)

### chore
- replace `if` with `with`
- use named ports in services
- remove `metrics.path` (user should not change that value)
- remove not used `ingress.hosts.paths.backend` from `values.yaml`

### CODE REVIEW

File: charts/server/templates/statefulset.yaml [hunk 1]
```diff
@@ -94,6 +94,10 @@ spec:
             - secretRef:
                 name: {{ . }}
             {{- end }}
+            {{- if .Values.woodpecker.metrics }}
```
Comment (pat-s) [inline charts/server/templates/statefulset.yaml:97]: I think it would be better if this would be a top-level option. I see no need to nest it under `woodpecker`.

File: charts/server/templates/servicemonitor.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{{- if and .Values.prometheus.servicemonitor.enabled ( .Capabilities.APIVersions.Has "monitoring.coreos.com/v1" ) }}
```
Comment (pat-s) [inline charts/server/templates/servicemonitor.yaml:1]: This also requires `metrics: true` AFAICS - so I guess adding it as a condition here would be good.

File: charts/server/values.yaml [hunk 3]
```diff
@@ -41,6 +41,34 @@ extraSecretNamesForEnvFrom:
   # - woodpecker-github-secret
   - woodpecker-secret
 
+woodpecker:
+  # -- enable metrics in woodpecker (and create secret with random token)
+  metrics: false
```
Comment (pat-s) [inline charts/server/values.yaml:46]: Should there be a comment that `metrics: true` is required for prometheus and grafana to work?

**Review by pat-s — COMMENTED** [review]
Thanks! See comments inline.

A short intro/example in the README would be great - most know Prometheus but it would be good to a short bit of information what this chart actually provides after this PR.

And please remove `charts/server/grafana_dashboards/.gitkeep` - it seemed to have slipped in by accident?

**Comment by genofire** [comment]
`charts/server/grafana_dashboards/.gitkeep` was created, because otherwise git does not create this directory `charts/server/grafana_dashboards/`, because there are not yet any files inside.


Of course i could delete the support of grafana_dashboards (and prometheus-rules, because we did not have any dashboards (and rules) yet.

or i need

File: charts/server/templates/statefulset.yaml [hunk 4]
```diff
@@ -72,14 +72,14 @@ spec:
...
             {{- toYaml .Values.resources | nindent 12 }}
-          {{- if .Values.persistentVolume.enabled }}
           volumeMounts:
-          - name: data
-            mountPath: {{ .Values.persistentVolume.mountPath }}
-          {{- end }}
-          {{- if .Values.extraVolumeMounts }}
-            {{- toYaml .Values.extraVolumeMounts | nindent 12 }}
-          {{- end }}
+            {{- if .Values.persistentVolume.enabled }}
```
Comment (genofire) [inline charts/server/templates/statefulset.yaml:76]: just a small finding -> allow extraVolumeMounts without enable default persistentVolume

File: charts/server/templates/pod-monitor.yaml [hunk 5]
```diff
@@ -1,16 +1,19 @@
-{{- if .Values.metrics.enabled }}
+{{- if and .Values.metrics.enabled .Values.prometheus.podmonitor.enabled ( .Capabilities.APIVersions.Has "monitoring.coreos.com/v1" ) }}
```
Comment (genofire) [inline charts/server/templates/pod-monitor.yaml:1]: this is a breaking change (that podmonitor must be enabled either and/or Capabilities-Check) ... ala semver

File: charts/server/templates/statefulset.yaml (hunk 4 above)
Comment (pat-s) [reply charts/server/templates/statefulset.yaml:76]: Thanks! Let's outsource this one to a new issue/PR

Comment (pat-s) [reply charts/server/templates/statefulset.yaml:76]: Actually it looks good in `main`: https://github.com/woodpecker-ci/helm/blob/6cfa8bd23c118c4f49379e8269a3c8076198c045/charts/server/templates/statefulset.yaml#L75-L79

File: charts/server/templates/statefulset.yaml [hunk 6]
```diff
@@ -106,9 +106,9 @@ spec:
       tolerations:
         {{- toYaml . | nindent 8 }}
       {{- end }}
+      {{- with .Values.extraVolumes }}
       volumes:
-      {{- if .Values.extraVolumes }}
-        {{- toYaml .Values.extraVolumes | nindent 8 }}
+        {{- toYaml . | nindent 8 }}
```
Comment (pat-s) [inline charts/server/templates/statefulset.yaml:111]: Is there a practical different using the `with` notation instead of the existing one?

File: charts/server/templates/pod-monitor.yaml [hunk 7]
```diff
@@ -1,16 +1,19 @@
...
+    {{- toYaml . | nindent 4}}
+    {{- end }}
 spec:
   selector:
     matchLabels:
-    {{- include "woodpecker-server.selectorLabels" . | nindent 6 }}
+      {{- include "woodpecker-server.selectorLabels" . | nindent 6 }}
   podMetricsEndpoints:
-  - port: metrics
-    interval: {{ .Values.metrics.interval }}
```
Comment (pat-s) [inline charts/server/templates/pod-monitor.yaml:14]: Should we keep the `interval` parameter? What is the default when unsetting it?

File: charts/server/values.yaml [hunk 8]
```diff
@@ -164,7 +192,6 @@ dnsConfig: {}
 
 # -- Enabled internal metrics endpoint.
 metrics:
+  # -- enable metrics in woodpecker
   enabled: false
   port: 9001
-  interval: 60s
-  path: /metrics
```
Comment (pat-s) [inline charts/server/values.yaml:170]: I'm not sure if WP supports changing the metrics path but in any case, we could leave the param as is? (Just to have less breaking changes)

File: charts/server/templates/statefulset.yaml (hunk 6 above)
Comment (genofire) [reply charts/server/templates/statefulset.yaml:111]: first you do not need to use the howl path again (here in toYaml), so it is easier to read.

secound, `volumes:` is not need without `extraVolumes`

File: charts/server/templates/pod-monitor.yaml (hunk 7 above)
Comment (genofire) [reply charts/server/templates/pod-monitor.yaml:14]: i just make th indent more beautiful - we keep interval and put that under `prometheus.podmonitor.interval` in `values.yaml`

File: charts/server/values.yaml (hunk 8 above)
Comment (genofire) [reply charts/server/values.yaml:170]: if the application change it, the helmchart should be change also (and no user of the helmchart).

or do you plan to make the path it configurable in the woodpecker-server itself

File: charts/server/templates/statefulset.yaml (hunk 4 above)
Comment (genofire) [reply charts/server/templates/statefulset.yaml:76]: No it does not ...

if `Values.persistentVolume.enabled: false` and `extraVolumeMounts` is set, it will generate invalid yaml

File: charts/server/values.yaml (hunk 8 above)
Comment (genofire) [reply charts/server/values.yaml:170]: but yes, we could let **interval** still be there.
(i personal prefer a clean seperation of metric enabled and usage of a PodMonitor).

so if you like it there i revert it

File: charts/server/templates/pod-monitor.yaml (hunk 7 above)
Comment (genofire) [reply charts/server/templates/pod-monitor.yaml:14]: you are correct, i make it optional

the default value of an prometheus server is 1m, this global scrape_interval could be changed, and that value could be overwritten by this Monitor :)

**Comment by genofire** [comment]
wow for the fast review (multiply times a day, my last just get a "please rebase" after four mouth).

i do not know, how you create / generate changelogs. It does not look like convention-commit based ... so i write some in the description.

File: charts/server/templates/statefulset.yaml (hunk 4 above)
Comment (pat-s) [reply charts/server/templates/statefulset.yaml:76]: Ahh that's what you mean. I second that. I thought until now you were referring to `extraVolumeMounts` being conditioned on `persistentVolume.enabled`.

File: charts/server/values.yaml (hunk 8 above)
Comment (pat-s) [reply charts/server/values.yaml:170]: > or do you plan to make the path it configurable in the woodpecker-server itself

I don't think so but I am just trying to ask "smart" questions :) I also didn't put the value there initially.

> (i personal prefer a clean seperation of metric enabled and usage of a PodMonitor).

Clean separations are always good but if this is resulting in a breaking change anyhow, we could also make more "breaking" changes.

I'll let it to you where to place `interval` - but if you move it, please document it -> i.e. we need some upgrade notes for this PR anyhow :)

**Comment by pat-s** [comment]
> wow for the fast review (multiply times a day, my last just get a "please rebase" after four mouth).

I am just contributing since some months - I wasn't (really) active during your last PR.

> i do not know, how you create / generate changelogs.

It's done automatically by a release bot - which scrapes the commit message/PR titles and merges them with the labels assigned.

But for this one we should add a upgrade changelog in the README - so it would be good to do it there, i.e. in `README-main.md.gotmpl`.

File: charts/server/values.yaml (hunk 8 above)
Comment (genofire) [reply charts/server/values.yaml:170]: where, in which file should i document that breaking change?

**Comment by genofire** [comment]
So should i split this PR in several PRs to get a nice CHANGELOG like in my description.

**Comment by pat-s** [comment]
Can you please add a new section to the README (below "Versioning") as follows

```
## Upgrading

<details>

<summary>To 1.0.0</summary>

YOUR CONTENT HERE

</details>
```

and replace `YOUR CONTENT HERE` with the user-facing upgrade notes? That would be awesome. No need for another PR, can go in here.

No features, just what users need to change if they were already using parts of the old implementation. Features etc. go into the release changelog then.

**Comment by genofire** [comment]
done, i have also improve the with/if on a place in the agent (but not change any logic)

File: Chart.lock [hunk 9]
```diff
@@ -1,9 +1,9 @@
 dependencies:
 - name: server
   repository: ""
-  version: 0.2.0
+  version: 0.3.0
```
Comment (pat-s) [inline Chart.lock:4]: Version bumps like these are done automatically during the release - no need to do it here :)

File: README-main.md.gotmpl [hunk 10]
```diff
@@ -55,6 +55,16 @@ resource "helm_release" "woodpecker" {
   timeout          = 90 # optional
 ```
 
+## Upgrading
+
+<details>
+
+<summary>To 1.0.0</summary>
+* split `metrics.enabled` and the PodMonitor deployment to `prometheus.podmonitor.enabled`
+    * move `metrics.interval` to `prometheus.podmonitor.interval`
```
Comment (pat-s) [inline README-main.md.gotmpl:64]:
```suggestion
<summary>To 1.0.0</summary>

* split `metrics.enabled` and the PodMonitor deployment to `prometheus.podmonitor.enabled`
* move `metrics.interval` to `prometheus.podmonitor.interval`
```

**Review by pat-s — COMMENTED** [review]
Thanks! A few last comments:

- Please revert the version upgrades (they are done during the release and are also inconsistent in this PR)
- Please avoid force pushes for future PRs as this makes it hard to see what changed (I remember you did this also in other/previous PRs - it's quite uncommon to do so in a regular PR workflow)
- Please merge `main` cleanly, then the `dind` removal changes should not appear in the PR

Besides: thanks for contributing 👍

**Comment by genofire** [comment]
1. done
2. sorry, i understand your workflow (with squashing on merge) now
3. i use the main branch ... should i go back in commits?

File: Chart.lock (hunk 9 above)
Comment (pat-s) [reply Chart.lock:4]: @genofire If you want to bump the subchart, this should be 1.0.0 due to the breaking changes.

File: charts/server/Chart.yaml [hunk 11]
```diff
@@ -2,7 +2,7 @@ apiVersion: v2
 name: server
 description: A Helm chart for the Woodpecker server
 type: application
-version: 0.2.0
+version: 0.3.0
```
Comment (pat-s) [inline charts/server/Chart.yaml:5]: @genofire If you want to bump the subchart, this should be 1.0.0 due to the breaking changes.

File: README-main.md.gotmpl [hunk 12]
```diff
@@ -55,6 +55,17 @@ resource "helm_release" "woodpecker" {
...
+
+<details>
+
+<summary>To 1.0.0</summary>
+
+* split `metrics.enabled` and the PodMonitor deployment to `prometheus.podmonitor.enabled`
+* move `metrics.interval` to `prometheus.podmonitor.interval`
+
+</details>
+
```
Comment (pat-s) [inline README-main.md.gotmpl:68]: These changes should also go to `charts/server/README.md`.

File: charts/server/values.yaml [hunk 13]
```diff
@@ -107,7 +136,7 @@ service:
   # -- The type of the service
   type: ClusterIP
   # -- The port of the service
-  port: &servicePort 80
```
Comment (pat-s) [inline charts/server/values.yaml:110]: Was this one undefined?

File: charts/server/values.yaml [hunk 14]
```diff
@@ -128,9 +157,6 @@ ingress:
     - host: chart-example.local
       paths:
         - path: /
-          backend:
-            serviceName: chart-example.local
-            servicePort: *servicePort
```
Comment (pat-s) [inline charts/server/values.yaml:133]: Why this change?

**Review by pat-s — COMMENTED** [review]
Almost there :)

File: charts/server/values.yaml (hunk 14 above)
Comment (genofire) [reply charts/server/values.yaml:133]: This values has no effect here / are deadcode.
The needed values are hardcoded (like it is needed).

https://github.com/woodpecker-ci/helm/blob/3c4c246b89c4e4849b62f62dab2b8a4c36521be6/charts/server/templates/ingress.yaml#L34-L41

File: README-main.md.gotmpl (hunk 12 above)
Comment (pat-s) [reply README-main.md.gotmpl:68]: > okay, i improve also the make docs, that we do not need the Makefile anymore (and everywhere we god the templates README.md.gotmpl)
so you just need helm-docs to run.

It was like this on purpose as the subcharts and the main chart have different templates (first `helm docs` renders all READMES and them overwrites the main README again).

Can we please revert and focus on the PR at hand? 😅

File: charts/server/values.yaml (hunk 14 above)
Comment (pat-s) [reply charts/server/values.yaml:133]: Thanks, I never looked at them in detail and always overwrite them anyhow. I am fine with removing them.

File: README-main.md.gotmpl (hunk 12 above)
Comment (genofire) [reply README-main.md.gotmpl:68]: i revert it.

(i solve your described problem with to put in every chart folder an template ...)

Comment (pat-s) [reply README-main.md.gotmpl:68]: Appreciated but discussed this internally already and it's not that easy in the current state. There's also #101 which will most likely tackle this. Anyhow, it will be solved at some point :)

**Review by pat-s — APPROVED** [review]
Thanks again! 🚀

**Comment by genofire** [comment]
how does it released?

**Comment by pat-s** [comment]
When WP 2.0 gets released
